runtime/runtime/src/prefetch.rs

```diff
@@ -0,0 +1,303 @@
+//! implementations could still negatively impact performance in corner cases.
+//! For example, if a prefetcher looks too far ahead, it may end up evicting
+//! data from LRU caches that is still required. A prefetcher could also use too
+//! much of the available I/O bandwidth and therefore slow down the main
+//! thread's I/O requests.
+//!
+//! The design goal is therefore that the prefetcher affects the performance of
+//! the main thread as little as possible. This is partially accomplished by
+//! treating the main thread's caches as read-only. All prefetched data is
+//! inserted in a separate prefetcher landing area. The main thread can check
+//! that space and move it to its own LRU caches if it finds the data there.
+//! But the prefetcher never directly modifies the main thread's caches.
+//! Reading from the LRU cache only makes some values stay longer in the cache.
+//! Presumably this is only positive, as the prefetcher knows the value will be
+//! used again.
+//!
+//! Another important measure is to limit the prefetcher in how far ahead it
+//! should prefetch, how much bandwidth it consumes, and how much memory it
+//! uses. We achieve this with a bounded queue for incoming requests, limiting
+//! the number of IO threads, and memory checks before staring new DB requests
+//! in the prefetcher. Implementation details for most limits are in
+//! `core/store/src/trie/prefetching_trie_storage.rs`
+
+use near_primitives::receipt::{Receipt, ReceiptEnum};
+use near_primitives::transaction::SignedTransaction;
+use near_primitives::trie_key::TrieKey;
+use near_store::{PrefetchApi, Trie};
+use std::rc::Rc;
+use tracing::debug;
+
+/// How many threads will be prefetching data, without the scheduler thread.
+/// Because the storage driver is blocking, there is only one request per thread
+/// at the time.
+const NUM_IO_THREADS: usize = 8;
+
+/// Transaction runtime view of the prefetching subsystem.
+pub(crate) struct TriePrefetcher {
+    prefetch_api: PrefetchApi,
+    #[cfg(test)]
+    handles: Vec<std::thread::JoinHandle<()>>,
+}
+
+impl TriePrefetcher {
+    pub(crate) fn new(trie: Rc<Trie>) -> Option<Self> {
+        if let Some(caching_storage) = trie.storage.as_caching_storage() {
+            let trie_root = trie.get_root().clone();
+            let parent = caching_storage.clone();
+            let prefetch_api = PrefetchApi::new(&parent);
+            #[cfg(test)]
+            let mut handles = vec![];
+            for _ in 0..NUM_IO_THREADS {
+                let _handle = prefetch_api.start_io_thread(parent, trie_root);
+                #[cfg(test)]
+                handles.push(_handle);
+            }
+            Some(Self {
+                prefetch_api,
+                #[cfg(test)]
+                handles,
+            })
+        } else {
+            None
+        }
+    }
+
+    /// Start prefetching data for processing the receipts.
+    ///
+    /// Returns an error if the prefetching queue is full.
```

Comment:
Thinking about it now, AFAICT the queue is already semantically bounded by the number of incoming transactions and receipts, which are all already held in memory, and gets cleared at the end of each block.

So I'm starting to wonder whether the queue should not just be unbounded, as sure it'll increase the memory use constant a bit, but this way it avoids the situation where data that could have been prefetched was not due to the queue being bounded?

Reply:
With the follow-up PR that is in the works, the amount of requests per receipts will no longer be bounded by a reasonable constant. So, while in the context of only prefetching accounts and access keys your argument makes a lot of sense, the more general prefetching I want to introduce will potentially read many items per receipts.
To give a concrete example, imagine 10 receipts are ready. Each of them is a IO heavy function call that we can somehow prefetch. Each will end up using 299 Tgas, so in reality we will only process 4 of them. Fetching data for all 10 receipts would be too much in this case.
Using a bounded queue is a way to prevent that we will prefetch more data than we can feasibly fetch within one block, without the need to know or predict how many of the receipts will be executed this block.